File: src/typing/destructuring.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

(* Destructuring visitor for tree-shaped patterns, parameteric over an action f
   to perform at the leaves. A type for the pattern is passed, which is taken
   apart as the visitor goes deeper. *)

(** TODO currently type annotations internal to patterns get parsed but not
  * checked. We should either update this to give users a warning that internal
  * annotations aren't checked, or update this to check internal annotations.
  *)

module Ast = Flow_ast
module Tast_utils = Typed_ast_utils
open Reason
open Type
open TypeUtil

module Make (Statement : Statement_sig.S) : Destructuring_sig.S = struct
  type state = {
    has_parent: bool;
    init: (ALoc.t, ALoc.t) Flow_ast.Expression.t option;
    default: Type.t Default.t option;
  }

  type callback =
    use_op:Type.use_op -> name_loc:ALoc.t -> string -> Type.t Default.t option -> Type.t -> Type.t

  let empty ?init ?default ~annot:_ _current = { has_parent = false; init; default }

  let pattern_default cx (acc : state) = function
    | None -> (acc, None)
    | Some e ->
      let default = acc.default in
      let (((_, t), _) as e) = Statement.expression cx e in
      let default = Some (Default.expr ?default t) in
      let acc = { acc with default } in
      (acc, Some e)

  let array_element cx acc i loc =
    let { init; default; _ } = acc in
    let key =
      DefT (mk_reason RNumber loc, bogus_trust (), NumT (Literal (None, (float i, string_of_int i))))
    in
    let reason = mk_reason (RCustom (Utils_js.spf "element %d" i)) loc in
    let init =
      Base.Option.map init ~f:(fun init ->
          ( loc,
            let open Ast.Expression in
            Member
              {
                Member._object = init;
                property =
                  Member.PropertyExpression
                    ( loc,
                      Ast.Expression.Literal
                        {
                          Ast.Literal.value = Ast.Literal.Number (float i);
                          raw = string_of_int i;
                          comments = None;
                        }
                    );
                comments = None;
              }
          )
      )
    in
    let refinement =
      Base.Option.bind init ~f:(fun init -> Refinement.get ~allow_optional:true cx init loc)
    in
    let has_parent = Base.Option.is_none refinement in
    let default = Base.Option.map default ~f:(Default.elem key reason) in
    { has_parent; init; default }

  let array_rest_element (acc : state) i loc =
    let default = acc.default in
    let reason = mk_reason RArrayPatternRestProp loc in
    let default = Base.Option.map default ~f:(Default.arr_rest i reason) in
    { acc with has_parent = true; default }

  let object_named_property ~has_default ~parent_loc cx acc loc x comments =
    let { init; default; _ } = acc in
    let reason = mk_reason (RProperty (Some (OrdinaryName x))) loc in
    let init =
      Base.Option.map init ~f:(fun init ->
          ( loc,
            let open Ast.Expression in
            Member
              {
                Member._object = init;
                property = Member.PropertyIdentifier (loc, { Ast.Identifier.name = x; comments });
                comments = None;
              }
          )
      )
    in
    let refinement =
      Base.Option.bind init ~f:(fun init -> Refinement.get ~allow_optional:true cx init loc)
    in
    let default =
      Base.Option.map default ~f:(fun default ->
          let d = Default.prop x reason has_default default in
          if has_default then
            Default.default reason d
          else
            d
      )
    in
    let parent_loc =
      match refinement with
      | Some _ -> None
      | None -> Some parent_loc
    in
    let has_parent = Base.Option.is_some parent_loc in
    let () =
      match parent_loc with
      | None -> () (* TODO: get-def when object property is refined *)
      | Some l ->
        (*
         * We are within a destructuring pattern and a `get-def` on this identifier should
         * point at the "def" of the original property. To accompish this, we emit the type
         * of the parent pattern so that get-def can dive in to that type and extract the
         * location of the "def" of this property.
         *)
        let t = Env.find_write cx Env_api.PatternLoc (mk_reason RDestructuring l) in
        Type_inference_hooks_js.dispatch_lval_hook cx x loc (Type_inference_hooks_js.Parent t)
    in
    { has_parent; init; default }

  let object_computed_property cx acc e =
    let { init; default; _ } = acc in
    let (((loc, t), _) as e') = Statement.expression cx e in
    let reason = mk_reason (RProperty None) loc in
    let init =
      Base.Option.map init ~f:(fun init ->
          ( loc,
            Ast.Expression.(
              Member Member.{ _object = init; property = PropertyExpression e; comments = None }
            )
          )
      )
    in
    let default = Base.Option.map default ~f:(Default.elem t reason) in
    ({ has_parent = true; init; default }, e')

  let object_rest_property (acc : state) xs loc =
    let default = acc.default in
    let reason = mk_reason RObjectPatternRestProp loc in
    let default = Base.Option.map default ~f:(Default.obj_rest xs reason) in
    { acc with has_parent = true; default }

  let object_property
      cx
      ~has_default
      ~parent_loc
      ~current
      (acc : state)
      xs
      (key : (ALoc.t, ALoc.t) Ast.Pattern.Object.Property.key) :
      state * string list * (ALoc.t, ALoc.t * Type.t) Ast.Pattern.Object.Property.key =
    let open Ast.Pattern.Object in
    match key with
    | Property.Identifier (loc, { Ast.Identifier.name = x; comments }) ->
      let acc = object_named_property ~has_default ~parent_loc cx acc loc x comments in
      (acc, x :: xs, Property.Identifier ((loc, current), { Ast.Identifier.name = x; comments }))
    | Property.Literal (loc, ({ Ast.Literal.value = Ast.Literal.String x; _ } as lit)) ->
      let acc = object_named_property ~has_default ~parent_loc cx acc loc x None in
      (acc, x :: xs, Property.Literal (loc, lit))
    | Property.Computed (loc, { Ast.ComputedKey.expression; comments }) ->
      let (acc, e) = object_computed_property cx acc expression in
      (acc, xs, Property.Computed (loc, { Ast.ComputedKey.expression = e; comments }))
    | Property.Literal (loc, _) ->
      Flow_js.add_output
        cx
        Error_message.(EUnsupportedSyntax (loc, DestructuringObjectPropertyLiteralNonString));
      (acc, xs, Tast_utils.error_mapper#pattern_object_property_key key)

  let identifier cx ~f acc name_loc name =
    let { has_parent; init; default } = acc in
    let () =
      if has_parent then
        (* If there was a parent pattern, we already dispatched the hook if relevant. *)
        ()
      else
        (*
         * If there was no parent_pattern, we must not be within a destructuring
         * pattern and a `get-def` on this identifier should point at the
         * location where the binding is introduced.
         *)
        Type_inference_hooks_js.dispatch_lval_hook cx name name_loc Type_inference_hooks_js.Id
    in
    let reason = mk_reason (RIdentifier (OrdinaryName name)) name_loc in
    let current =
      mod_reason_of_t
        (update_desc_reason (function
            | RDefaultValue
            | RArrayPatternRestProp
            | RObjectPatternRestProp ->
              RIdentifier (OrdinaryName name)
            | desc -> desc
            )
            )
        (Env.find_write cx Env_api.OrdinaryNameLoc reason)
    in
    let use_op =
      Op
        (AssignVar
           {
             var = Some reason;
             init =
               (match (default, init) with
               | (Some (Default.Expr t), _) -> reason_of_t t
               | (_, Some init) -> mk_expression_reason init
               | _ -> reason_of_t current);
           }
        )
    in
    f ~use_op ~name_loc name default current

  let current_type cx (loc, p) =
    match p with
    | Ast.Pattern.Identifier
        { Ast.Pattern.Identifier.name = (name_loc, { Ast.Identifier.name; _ }); _ } ->
      Env.find_write
        cx
        Env_api.OrdinaryNameLoc
        (mk_reason (RIdentifier (OrdinaryName name)) name_loc)
    | Ast.Pattern.Expression _ ->
      (* Expression in pattern destructuring is unsupported syntax,
         so we shouldn't read the environment. *)
      AnyT.untyped (mk_reason RDestructuring loc)
    | _ -> Env.find_write cx Env_api.PatternLoc (mk_reason RDestructuring loc)

  let rec pattern cx ~(f : callback) acc (loc, p) =
    let check_for_invalid_annot annot =
      match (acc.has_parent, annot) with
      | (true, Ast.Type.Available (loc, _)) ->
        Flow_js.add_output cx Error_message.(EUnsupportedSyntax (loc, AnnotationInsideDestructuring))
      | _ -> ()
    in
    let open Ast.Pattern in
    ( (loc, current_type cx (loc, p)),
      match p with
      | Array { Array.elements; annot; comments } ->
        check_for_invalid_annot annot;
        let elements = array_elements cx ~f acc elements in
        let annot = Tast_utils.unimplemented_mapper#type_annotation_hint annot in
        Array { Array.elements; annot; comments }
      | Object { Object.properties; annot; comments } ->
        check_for_invalid_annot annot;
        let properties = object_properties cx ~f ~parent_loc:loc acc properties in
        let annot = Tast_utils.unimplemented_mapper#type_annotation_hint annot in
        Object { Object.properties; annot; comments }
      | Identifier { Identifier.name = id; optional; annot } ->
        let (id_loc, { Ast.Identifier.name; comments }) = id in
        check_for_invalid_annot annot;
        let annot = Tast_utils.unimplemented_mapper#type_annotation_hint annot in
        let id_ty = identifier cx ~f acc id_loc name in
        let id = ((id_loc, id_ty), { Ast.Identifier.name; comments }) in
        Identifier { Identifier.name = id; optional; annot }
      | Expression e ->
        Flow_js.add_output
          cx
          Error_message.(EUnsupportedSyntax (loc, DestructuringExpressionPattern));
        Expression (Tast_utils.error_mapper#expression e)
    )

  and array_elements cx ~f acc =
    let open Ast.Pattern.Array in
    Base.List.mapi ~f:(fun i -> function
      | Hole loc -> Hole loc
      | Element (loc, { Element.argument = p; default = d }) ->
        let acc = array_element cx acc i loc in
        let (acc, d) = pattern_default cx acc d in
        let p = pattern cx ~f acc p in
        Element (loc, { Element.argument = p; default = d })
      | RestElement (loc, { Ast.Pattern.RestElement.argument = (arg_loc, _) as p; comments }) ->
        let acc = array_rest_element acc i arg_loc in
        let p = pattern cx ~f acc p in
        RestElement (loc, { Ast.Pattern.RestElement.argument = p; comments })
    )

  and object_properties =
    let open Ast.Pattern.Object in
    let prop cx ~f ~parent_loc acc xs p =
      match p with
      | Property (loc, { Property.key; pattern = p; default = d; shorthand }) ->
        let has_default = d <> None in
        let current = current_type cx p in
        let (acc, xs, key) = object_property cx ~has_default ~parent_loc ~current acc xs key in
        let (acc, d) = pattern_default cx acc d in
        let p = pattern cx ~f acc p in
        (xs, Property (loc, { Property.key; pattern = p; default = d; shorthand }))
      | RestElement (loc, { Ast.Pattern.RestElement.argument = (arg_loc, _) as p; comments }) ->
        let acc = object_rest_property acc xs arg_loc in
        let p = pattern cx ~f acc p in
        (xs, RestElement (loc, { Ast.Pattern.RestElement.argument = p; comments }))
    in
    let rec loop cx ~f ~parent_loc acc xs rev_ps = function
      | [] -> List.rev rev_ps
      | p :: ps ->
        let (xs, p) = prop cx ~f ~parent_loc acc xs p in
        loop cx ~f ~parent_loc acc xs (p :: rev_ps) ps
    in
    (fun cx ~f ~parent_loc acc ps -> loop cx ~f ~parent_loc acc [] [] ps)

  let type_of_pattern (_, p) =
    let open Ast.Pattern in
    match p with
    | Array { Array.annot; _ }
    | Object { Object.annot; _ }
    | Identifier { Identifier.annot; _ } ->
      annot
    | _ -> Ast.Type.Missing ALoc.none

  (* instantiate pattern visitor for assignments *)
  let assignment cx rhs_t init =
    let acc = empty ~init ~annot:false rhs_t in
    let f ~use_op ~name_loc name _default t =
      (* TODO destructuring+defaults unsupported in assignment expressions *)
      ignore Env.(set_var cx ~use_op name t name_loc);
      Env.constraining_type ~default:t cx name_loc
    in
    pattern cx ~f acc
end

File: rust_port/crates/flow_typing_statement/src/destructuring.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! Destructuring visitor for tree-shaped patterns, parameteric over an action f
//! to perform at the leaves. A type for the pattern is passed, which is taken
//! apart as the visitor goes deeper.

use std::sync::Arc;

use dupe::Dupe;
use flow_aloc::ALoc;
use flow_common::reason::Name;
use flow_common::reason::Reason;
use flow_common::reason::VirtualReasonDesc;
use flow_common::reason::mk_reason;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_env_builder::env_api::DefLocType;
use flow_parser::ast;
use flow_parser::ast::expression;
use flow_parser::ast::pattern;
use flow_parser::ast_utils;
use flow_parser::loc_sig::LocSig;
use flow_parser::polymorphic_ast_mapper;
use flow_parser::polymorphic_ast_mapper::LocMapper;
use flow_typing_context::Context;
use flow_typing_default as default;
use flow_typing_errors::error_message::ErrorMessage;
use flow_typing_errors::intermediate_error_types::UnsupportedSyntax;
use flow_typing_flow_js::flow_js;
use flow_typing_type::type_;
use flow_typing_type::type_::DefT;
use flow_typing_type::type_::DefTInner;
use flow_typing_type::type_::NumberLiteral;
use flow_typing_type::type_::RootUseOp;
use flow_typing_type::type_::Type;
use flow_typing_type::type_::TypeInner;
use flow_typing_type::type_::UseOp;
use flow_typing_type::type_::any_t;
use flow_typing_type::type_util;
use flow_typing_utils::abnormal::AbnormalControlFlow;
use flow_typing_utils::type_env;
use flow_typing_utils::typed_ast_utils;
use flow_typing_utils::typed_ast_utils::ErrorMapper;

use crate::statement;
pub struct State {
    has_parent: bool,
    init: Option<expression::Expression<ALoc, ALoc>>,
    default: Option<default::Default<Type>>,
}

pub type Callback<'a> = dyn Fn(
        /*use_op:*/ &UseOp,
        /*name_loc:*/ ALoc,
        /*name:*/ &FlowSmolStr,
        /*default:*/ Option<&default::Default<Type>>,
        /*current:*/ Type,
    ) -> Type
    + 'a;

pub fn empty(
    init: Option<expression::Expression<ALoc, ALoc>>,
    default: Option<default::Default<Type>>,
) -> State {
    State {
        has_parent: false,
        init,
        default,
    }
}

fn pattern_default<'a>(
    cx: &Context<'a>,
    acc: &mut State,
    default_expr: Option<&expression::Expression<ALoc, ALoc>>,
) -> Result<Option<expression::Expression<ALoc, (ALoc, Type)>>, AbnormalControlFlow> {
    match default_expr {
        None => Ok(None),
        Some(e) => {
            let prev_default = acc.default.take();
            let e = statement::expression(None, None, None, cx, e)?;
            let t = e.loc().1.dupe();
            let new_default = Some(default::expr(t, prev_default));
            acc.default = new_default;
            Ok(Some(e))
        }
    }
}

fn array_element<'a>(cx: &Context<'a>, acc: &State, i: i32, loc: ALoc) -> State {
    let init = &acc.init;
    let default_val = &acc.default;
    let key = Type::new(TypeInner::DefT(
        mk_reason(VirtualReasonDesc::RNumber, loc.dupe()),
        DefT::new(DefTInner::SingletonNumT {
            from_annot: false,
            value: NumberLiteral(i as f64, FlowSmolStr::new(i.to_string())),
        }),
    ));
    let reason = mk_reason(VirtualReasonDesc::RArrayNthElement(i), loc.dupe());
    let new_init = init.as_ref().map(|init| {
        expression::Expression::new(expression::ExpressionInner::Member {
            loc: loc.dupe(),
            inner: Arc::new(expression::Member {
                object: init.dupe(),
                property: expression::member::Property::PropertyExpression(
                    expression::Expression::new(expression::ExpressionInner::NumberLiteral {
                        loc: loc.dupe(),
                        inner: Arc::new(ast::NumberLiteral {
                            value: i as f64,
                            raw: FlowSmolStr::new(i.to_string()),
                            comments: None,
                        }),
                    }),
                ),
                comments: None,
            }),
        })
    });
    let refinement = new_init
        .as_ref()
        .and_then(|init| crate::refinement::get(true, cx, init, loc.dupe()));
    let has_parent = refinement.is_none();
    let new_default = default_val
        .as_ref()
        .map(|d| default::elem(key, reason, d.clone()));
    State {
        has_parent,
        init: new_init,
        default: new_default,
    }
}

fn array_rest_element(acc: &State, i: i32, loc: ALoc) -> State {
    let default_val = &acc.default;
    let reason = mk_reason(VirtualReasonDesc::RArrayPatternRestProp, loc.dupe());
    let new_default = default_val
        .as_ref()
        .map(|d| default::arr_rest(i, reason, d.clone()));
    State {
        has_parent: true,
        init: acc.init.dupe(),
        default: new_default,
    }
}

fn object_named_property<'a>(
    has_default: bool,
    _parent_loc: ALoc,
    cx: &Context<'a>,
    acc: &State,
    loc: ALoc,
    x: &FlowSmolStr,
    comments: Option<ast::Syntax<ALoc, ()>>,
) -> State {
    let init = &acc.init;
    let default_val = &acc.default;
    let reason = mk_reason(
        VirtualReasonDesc::RProperty(Some(Name::new(FlowSmolStr::new(x)))),
        loc.dupe(),
    );
    let new_init = init.as_ref().map(|init| {
        expression::Expression::new(expression::ExpressionInner::Member {
            loc: loc.dupe(),
            inner: Arc::new(expression::Member {
                object: init.dupe(),
                property: expression::member::Property::PropertyIdentifier(ast::Identifier::new(
                    ast::IdentifierInner {
                        loc: loc.dupe(),
                        name: x.dupe(),
                        comments,
                    },
                )),
                comments: None,
            }),
        })
    });
    let refinement = new_init
        .as_ref()
        .and_then(|init| crate::refinement::get(true, cx, init, loc.dupe()));
    let new_default = default_val.as_ref().map(|d| {
        let prop_d = default::prop(x.dupe(), reason.dupe(), has_default, d.clone());
        if has_default {
            default::default(reason.dupe(), prop_d)
        } else {
            prop_d
        }
    });
    let parent_loc_opt = match refinement {
        Some(_) => None,
        None => Some(_parent_loc),
    };
    let has_parent = parent_loc_opt.is_some();
    State {
        has_parent,
        init: new_init,
        default: new_default,
    }
}

fn object_computed_property<'a>(
    cx: &Context<'a>,
    acc: &State,
    e: &expression::Expression<ALoc, ALoc>,
) -> Result<(State, expression::Expression<ALoc, (ALoc, Type)>), AbnormalControlFlow> {
    let init = &acc.init;
    let default_val = &acc.default;
    let e_typed = statement::expression(None, None, None, cx, e)?;
    let (loc, t) = e_typed.loc().dupe();
    let reason = mk_reason(VirtualReasonDesc::RProperty(None), loc.dupe());
    let new_init = init.as_ref().map(|init| {
        expression::Expression::new(expression::ExpressionInner::Member {
            loc: loc.dupe(),
            inner: Arc::new(expression::Member {
                object: init.dupe(),
                property: expression::member::Property::PropertyExpression(e.dupe()),
                comments: None,
            }),
        })
    });
    let new_default = default_val
        .as_ref()
        .map(|d| default::elem(t, reason, d.clone()));
    Ok((
        State {
            has_parent: true,
            init: new_init,
            default: new_default,
        },
        e_typed,
    ))
}

fn object_rest_property(acc: &State, xs: &[FlowSmolStr], loc: ALoc) -> State {
    let default_val = &acc.default;
    let reason = mk_reason(VirtualReasonDesc::RObjectPatternRestProp, loc.dupe());
    let new_default = default_val
        .as_ref()
        .map(|d| default::obj_rest(xs, reason, d.clone()));
    State {
        has_parent: true,
        init: acc.init.dupe(),
        default: new_default,
    }
}

fn object_property<'a>(
    cx: &Context<'a>,
    has_default: bool,
    parent_loc: ALoc,
    current: Type,
    acc: &State,
    xs: &mut Vec<FlowSmolStr>,
    key: &pattern::object::Key<ALoc, ALoc>,
) -> Result<(State, pattern::object::Key<ALoc, (ALoc, Type)>), AbnormalControlFlow> {
    match key {
        pattern::object::Key::Identifier(id) => {
            let loc = id.loc.dupe();
            let x = id.name.dupe();
            let comments = id.comments.dupe();
            let new_acc = object_named_property(
                has_default,
                parent_loc,
                cx,
                acc,
                loc.dupe(),
                &x,
                comments.dupe(),
            );
            xs.push(x.dupe());
            Ok((
                new_acc,
                pattern::object::Key::Identifier(ast::Identifier::new(ast::IdentifierInner {
                    loc: (loc, current),
                    name: x,
                    comments,
                })),
            ))
        }
        pattern::object::Key::StringLiteral((loc, lit)) => {
            let x = &lit.value;
            let new_acc =
                object_named_property(has_default, parent_loc, cx, acc, loc.dupe(), x, None);
            xs.push(x.dupe());
            Ok((
                new_acc,
                pattern::object::Key::StringLiteral((loc.dupe(), lit.clone())),
            ))
        }
        pattern::object::Key::Computed(ck) => {
            let (new_acc, e) = object_computed_property(cx, acc, &ck.expression)?;
            Ok((
                new_acc,
                pattern::object::Key::Computed(ast::ComputedKey {
                    loc: ck.loc.dupe(),
                    expression: e,
                    comments: ck.comments.dupe(),
                }),
            ))
        }
        pattern::object::Key::NumberLiteral((loc, lit))
            if flow_common::js_number::is_float_safe_integer(lit.value) =>
        {
            let name = flow_common::js_number::ecma_string_of_float(lit.value);
            let name_smol = FlowSmolStr::new(&name);
            // let acc = object_named_property ~has_default ~parent_loc cx acc loc name comments in
            let new_acc = object_named_property(
                has_default,
                parent_loc,
                cx,
                acc,
                loc.dupe(),
                &name_smol,
                lit.comments.dupe(),
            );
            xs.push(name_smol);
            Ok((
                new_acc,
                pattern::object::Key::NumberLiteral((loc.dupe(), lit.clone())),
            ))
        }
        pattern::object::Key::NumberLiteral((loc, _))
        | pattern::object::Key::BigIntLiteral((loc, _)) => {
            flow_js::add_output_non_speculating(
                cx,
                ErrorMessage::EUnsupportedSyntax(Box::new((
                    loc.dupe(),
                    UnsupportedSyntax::DestructuringObjectPropertyInvalidLiteral,
                ))),
            );
            let mut mapper = ErrorMapper;
            let typed_key = match key {
                pattern::object::Key::NumberLiteral((loc, lit)) => {
                    pattern::object::Key::NumberLiteral((
                        {
                            let Ok(v) = mapper.on_loc_annot(loc);
                            v
                        },
                        lit.clone(),
                    ))
                }
                pattern::object::Key::BigIntLiteral((loc, lit)) => {
                    pattern::object::Key::BigIntLiteral((
                        {
                            let Ok(v) = mapper.on_loc_annot(loc);
                            v
                        },
                        lit.clone(),
                    ))
                }
                _ => unreachable!(),
            };
            Ok((
                State {
                    has_parent: acc.has_parent,
                    init: acc.init.dupe(),
                    default: acc.default.clone(),
                },
                typed_key,
            ))
        }
    }
}

fn identifier<'a>(
    cx: &Context<'a>,
    f: &Callback<'_>,
    acc: &State,
    name_loc: ALoc,
    name: &FlowSmolStr,
) -> Type {
    let default_val = &acc.default;
    let reason = mk_reason(
        VirtualReasonDesc::RIdentifier(Name::new(FlowSmolStr::new(name))),
        name_loc.dupe(),
    );
    let write_t = type_env::find_write(cx, DefLocType::OrdinaryNameLoc, reason.dupe());
    let current = type_util::mod_reason_of_t(
        &|r: Reason| {
            let desc = r.desc(false);
            let new_desc = match desc {
                VirtualReasonDesc::RDefaultValue
                | VirtualReasonDesc::RArrayPatternRestProp
                | VirtualReasonDesc::RObjectPatternRestProp => {
                    VirtualReasonDesc::RIdentifier(Name::new(FlowSmolStr::new(name)))
                }
                _ => desc.clone(),
            };
            r.replace_desc(new_desc)
        },
        &write_t,
    );
    let init_reason = match (default_val, &acc.init) {
        (Some(default::Default::Expr(t)), _) => type_util::reason_of_t(t).dupe(),
        (_, Some(init)) => flow_common::reason::mk_expression_reason(init),
        _ => type_util::reason_of_t(&current).dupe(),
    };
    let use_op = UseOp::Op(Arc::new(RootUseOp::AssignVar {
        var: Some(reason),
        init: init_reason,
    }));
    f(&use_op, name_loc, name, default_val.as_ref(), current)
}

fn current_type<'a>(cx: &Context<'a>, p: &pattern::Pattern<ALoc, ALoc>) -> Type {
    let loc = p.loc().dupe();
    match p {
        pattern::Pattern::Identifier { inner, .. } => {
            let name_loc = inner.name.loc.dupe();
            let name = &inner.name.name;
            type_env::find_write(
                cx,
                DefLocType::OrdinaryNameLoc,
                mk_reason(
                    VirtualReasonDesc::RIdentifier(Name::new(name.dupe())),
                    name_loc,
                ),
            )
        }
        pattern::Pattern::Expression { .. } => {
            // Expression in pattern destructuring is unsupported syntax,
            // so we shouldn't read the environment.
            any_t::untyped(mk_reason(VirtualReasonDesc::RDestructuring, loc))
        }
        _ => {
            if ast_utils::pattern_has_binding(p) {
                type_env::find_write(
                    cx,
                    DefLocType::PatternLoc,
                    mk_reason(VirtualReasonDesc::RDestructuring, loc),
                )
            } else {
                type_::unsoundness::at(type_::UnsoundnessKind::NonBindingPattern, loc)
            }
        }
    }
}

pub fn pattern<'a>(
    cx: &Context<'a>,
    f: &Callback<'_>,
    acc: &mut State,
    p: &pattern::Pattern<ALoc, ALoc>,
) -> Result<pattern::Pattern<ALoc, (ALoc, Type)>, AbnormalControlFlow> {
    let check_for_invalid_annot =
        |has_parent: bool, annot: &ast::types::AnnotationOrHint<ALoc, ALoc>| {
            if let (true, ast::types::AnnotationOrHint::Available(ann)) = (has_parent, annot) {
                flow_js::add_output_non_speculating(
                    cx,
                    ErrorMessage::EUnsupportedSyntax(Box::new((
                        ann.loc.dupe(),
                        UnsupportedSyntax::AnnotationInsideDestructuring,
                    ))),
                );
            }
        };

    let loc = p.loc().dupe();
    let ct = current_type(cx, p);

    match p {
        pattern::Pattern::Array { inner, .. } => {
            check_for_invalid_annot(acc.has_parent, &inner.annot);
            let elements = array_elements(cx, f, acc, &inner.elements)?;
            let mut mapper = typed_ast_utils::UnimplementedMapper;
            let Ok(annot) = polymorphic_ast_mapper::type_annotation_hint(&mut mapper, &inner.annot);
            Ok(pattern::Pattern::Array {
                loc: (loc.dupe(), ct),
                inner: pattern::Array {
                    elements: elements.into(),
                    annot,
                    optional: inner.optional,
                    comments: inner.comments.dupe(),
                }
                .into(),
            })
        }
        pattern::Pattern::Object { inner, .. } => {
            check_for_invalid_annot(acc.has_parent, &inner.annot);
            let properties = object_properties(cx, f, loc.dupe(), acc, &inner.properties)?;
            let mut mapper = typed_ast_utils::UnimplementedMapper;
            let Ok(annot) = polymorphic_ast_mapper::type_annotation_hint(&mut mapper, &inner.annot);
            Ok(pattern::Pattern::Object {
                loc: (loc.dupe(), ct),
                inner: pattern::Object {
                    properties: properties.into(),
                    annot,
                    optional: inner.optional,
                    comments: inner.comments.dupe(),
                }
                .into(),
            })
        }
        pattern::Pattern::Identifier { inner, .. } => {
            let id_loc = inner.name.loc.dupe();
            let name = &inner.name.name;
            let comments = &inner.name.comments;
            check_for_invalid_annot(acc.has_parent, &inner.annot);
            let mut mapper = typed_ast_utils::UnimplementedMapper;
            let Ok(annot) = polymorphic_ast_mapper::type_annotation_hint(&mut mapper, &inner.annot);
            let id_ty = identifier(cx, f, acc, id_loc.dupe(), name);
            let id = ast::Identifier::new(ast::IdentifierInner {
                loc: (id_loc, id_ty),
                name: name.dupe(),
                comments: comments.dupe(),
            });
            Ok(pattern::Pattern::Identifier {
                loc: (loc.dupe(), ct),
                inner: pattern::Identifier {
                    name: id,
                    optional: inner.optional,
                    annot,
                }
                .into(),
            })
        }
        pattern::Pattern::Expression { inner, .. } => {
            flow_js::add_output_non_speculating(
                cx,
                ErrorMessage::EUnsupportedSyntax(Box::new((
                    loc.dupe(),
                    UnsupportedSyntax::DestructuringExpressionPattern,
                ))),
            );
            let mut mapper = ErrorMapper;
            let Ok(typed_expr) = polymorphic_ast_mapper::expression(&mut mapper, inner);
            Ok(pattern::Pattern::Expression {
                loc: (loc.dupe(), ct),
                inner: typed_expr.into(),
            })
        }
    }
}

pub fn array_elements<'a>(
    cx: &Context<'a>,
    f: &Callback<'_>,
    acc: &mut State,
    elements: &[pattern::array::Element<ALoc, ALoc>],
) -> Result<Vec<pattern::array::Element<ALoc, (ALoc, Type)>>, AbnormalControlFlow> {
    elements
        .iter()
        .enumerate()
        .map(|(i, elem)| -> Result<_, AbnormalControlFlow> {
            match elem {
                pattern::array::Element::Hole(loc) => Ok(pattern::array::Element::Hole(loc.dupe())),
                pattern::array::Element::NormalElement(ne) => {
                    let loc = ne.loc.dupe();
                    let p = &ne.argument;
                    let d = &ne.default;
                    let mut elem_acc = array_element(cx, acc, i as i32, loc.dupe());
                    let typed_d = pattern_default(cx, &mut elem_acc, d.as_ref())?;
                    let typed_p = pattern(cx, f, &mut elem_acc, p)?;
                    Ok(pattern::array::Element::NormalElement(
                        pattern::array::NormalElement {
                            loc,
                            argument: typed_p,
                            default: typed_d,
                        },
                    ))
                }
                pattern::array::Element::RestElement(re) => {
                    let loc = re.loc.dupe();
                    let p = &re.argument;
                    let arg_loc = p.loc().dupe();
                    let mut rest_acc = array_rest_element(acc, i as i32, arg_loc);
                    let typed_p = pattern(cx, f, &mut rest_acc, p)?;
                    Ok(pattern::array::Element::RestElement(pattern::RestElement {
                        loc,
                        argument: typed_p,
                        comments: re.comments.dupe(),
                    }))
                }
            }
        })
        .collect::<Result<Vec<_>, _>>()
}

pub fn object_properties<'a>(
    cx: &Context<'a>,
    f: &Callback<'_>,
    parent_loc: ALoc,
    acc: &mut State,
    properties: &[pattern::object::Property<ALoc, ALoc>],
) -> Result<Vec<pattern::object::Property<ALoc, (ALoc, Type)>>, AbnormalControlFlow> {
    let mut xs = Vec::new();
    let mut result = Vec::new();

    for p in properties {
        match p {
            pattern::object::Property::NormalProperty(np) => {
                let loc = np.loc.dupe();
                let key = &np.key;
                let pat = &np.pattern;
                let d = &np.default;
                let shorthand = np.shorthand;
                let has_default = d.is_some();
                let current = current_type(cx, pat);
                let (mut prop_acc, typed_key) = object_property(
                    cx,
                    has_default,
                    parent_loc.dupe(),
                    current,
                    acc,
                    &mut xs,
                    key,
                )?;
                let typed_d = pattern_default(cx, &mut prop_acc, d.as_ref())?;
                let typed_p = pattern(cx, f, &mut prop_acc, pat)?;
                result.push(pattern::object::Property::NormalProperty(
                    pattern::object::NormalProperty {
                        loc,
                        key: typed_key,
                        pattern: typed_p,
                        default: typed_d,
                        shorthand,
                    },
                ));
            }
            pattern::object::Property::RestElement(re) => {
                let loc = re.loc.dupe();
                let pat = &re.argument;
                let arg_loc = pat.loc().dupe();
                let mut rest_acc = object_rest_property(acc, &xs, arg_loc);
                let typed_p = pattern(cx, f, &mut rest_acc, pat)?;
                result.push(pattern::object::Property::RestElement(
                    pattern::RestElement {
                        loc,
                        argument: typed_p,
                        comments: re.comments.dupe(),
                    },
                ));
            }
        }
    }

    Ok(result)
}

pub fn type_of_pattern(
    p: &pattern::Pattern<ALoc, ALoc>,
) -> ast::types::AnnotationOrHint<ALoc, ALoc> {
    match p {
        pattern::Pattern::Array { inner, .. } => inner.annot.clone(),
        pattern::Pattern::Object { inner, .. } => inner.annot.clone(),
        pattern::Pattern::Identifier { inner, .. } => inner.annot.clone(),
        _ => ast::types::AnnotationOrHint::Missing(ALoc::none()),
    }
}

/// instantiate pattern visitor for assignments
pub fn assignment<'a>(
    cx: &Context<'a>,
    init: expression::Expression<ALoc, ALoc>,
    p: &pattern::Pattern<ALoc, ALoc>,
) -> Result<pattern::Pattern<ALoc, (ALoc, Type)>, AbnormalControlFlow> {
    let mut acc = empty(Some(init), None);
    pattern(
        cx,
        &|use_op, name_loc, name, _default, t| {
            // TODO destructuring+defaults unsupported in assignment expressions
            type_env::set_var(cx, use_op, name, &t, name_loc.dupe());
            type_env::constraining_type(t, cx, name, name_loc)
        },
        &mut acc,
        p,
    )
}
